### ISSUE TITLE
JWS that are not JWT Proposal

### ISSUE THREAD

**Issue by OR13** [decentralized-identity/did-jose-extensions#2]
For JWS that are not JWT, `kid` SHOULD be a valid DID_URL of the form: `did:example:123#fragment`.

Verification is as follows, resolve `did:example:123#fragment` => did document (fragments are handled by clients not servers, just like http).... iterate the verification relationships until finding one which `id` matching `did:example:123#fragment`... convert to JWK if necessary, verify.

Warnings:

`fragment` cannot be trusted to be an RFC compliant thumbprint, and might be something like `key-0` or `pii-leaking-identifier-123`...

since a JWS is a vanilla signature, I would argue that its recommended that you only need to iterate `authentication`.. because if it were a credential the recommendation would be to use `assertionMethod` for compatibility with ld-proofs / so as not to make legacy support for JWS harm future facing support for LD Proofs.

**Comment by oed** [comment]
This does not exclude the use of a query param like `?version=1` right?

**Comment by OR13** [comment]
correct, that just a specific case of a DID URL with a DID Parameter...

We are concerned with the general form:

`did?params#fragment`

**Comment by oed** [comment]
> since a JWS is a vanilla signature, I would argue that its recommended that you only need to iterate authentication.. because if it were a credential the recommendation would be to use assertionMethod for compatibility with ld-proofs / so as not to make legacy support for JWS harm future facing support for LD Proofs.

Does this assume that the JWS is a credential? I definitely think there are a bunch of use cases where you'd want to sign something with a key that's not in `authentication`.

**Comment by OR13** [comment]
https://w3c.github.io/did-core/#verification-methods

TLDR... my recommendation is as follows:

JWTs that are VC => assertionMethod
JWTs that are VP => authentication
JWS => authentication

**Comment by oed** [comment]
Hey @OR13 What happened with the examples in the readme? Seems like they where changed to no longer include `did?params#fragment`. Is this an error?

**Comment by OR13** [comment]
No changes have been made in a while https://github.com/decentralized-identity/did-jose-extensions/commits/master

I never added any examples that used params, because I don't ever use them, but others may have.

**Comment by oed** [comment]
Ok, I thought that they where there at some point. I can make a PR to add them? We use them in https://eips.ethereum.org/EIPS/eip-2844 and related libraries.

**Comment by OR13** [comment]
yes, please add examples of JWS and DIDs that you use as issues, and then we probably need to take a first stab at formalizing this as a document and closing out the issues.

**Comment by letmaik** [comment]
Would it make sense to register `iss` as a JWS header parameter and use it for the DID (not DID URL) like in JWTs?

**Comment by oed** [comment]
We've been using the `kid` parameter for quite some time already.

**Comment by letmaik** [comment]
To be honest, I find it awkward to combine DID and key id in one field for a couple of reasons.
First, it requires every consumer to parse it apart whereas normally key id is considered opaque. It strangely feels like LD is creeping in here.
Second, `kid` is normally just a hint to help with key location within the scope of an issuer, whereas `iss` is a statement about who the issuer is. With JWTs, the issuer can be used in policies quite conveniently. I would feel very uneasy using `kid` in policies. 

Note that `iss` is already registered as [JOSE header parameter](https://www.iana.org/assignments/jose/jose.xhtml), though only for use in [JWEs in the context of the JWT spec](https://datatracker.ietf.org/doc/html/rfc7519#section-5.3).

Does it make sense in principle what I'm thinking, ignoring for now what it would take to actually pursue it?

**Comment by oed** [comment]
To me `kid` is used to point to the specific *Verification Method* that signed the JWS, not the *issuer* of the JWS. These are quite different imo.

**Comment by letmaik** [comment]
> 
> 
> To me `kid` is used to point to the specific _Verification Method_ that signed the JWS, not the _issuer_ of the JWS. These are quite different imo.

100% agree. I guess the contention point is whether there is value in trying to standardize a generic JWS method to identify the issuer.

**Comment by oed** [comment]
@letmaik I guess `iss` is already part of the JWT Verifiable Credential spec: https://www.w3.org/TR/vc-data-model/#jwt-encoding

**Comment by letmaik** [comment]
> 
> 
> @letmaik I guess `iss` is already part of the JWT Verifiable Credential spec: https://www.w3.org/TR/vc-data-model/#jwt-encoding

Right, but that's specifically for JWTs where `iss` is in the payload of the JWS. For generic JWS messages, you want to treat the payload as opaque, and so I was thinking whether it makes sense to establish the issuer concept in the JWS header as parameter.

**Comment by oed** [comment]
Do you have any use case for this in mind? 
@awoie might have some thoughts here.

**Comment by OR13** [comment]
related https://github.com/decentralized-identity/did-jose-extensions/issues/7

in short, `kid` being optional is really problematic, if you need to rely on it to identify keys.

so when working with DIDs, it better to use `verificationMethod.id` for `kid` and make `kid` always present in the JWS and never present in the `publicKeyJwk`... this will also help you if you need to support JSON-LD LD Proofs... which require an identifier for public key that can be dereferenced to the public key.... something which JOSE lacks... and which has held the door open to incompatibility issues across VC-JWT implementations.

**Comment by awoie** [comment]
- `iss` identifies already the issuer. 
- `kid` allows you to find the key to verify the JWS/JWT. i agree with @OR13 

We should probably say that `kid` (if DID URL) must correspond to `iss` (if DID) although technically, a VC can have a different `issuer` (DID) and `verificationMethod` (DID) even when using LD-Proofs.

**Comment by letmaik** [comment]
>     * `iss` identifies already the issuer.
>     * `kid` allows you to find the key to verify the JWS/JWT. i agree with @OR13

`iss` is not registered as JWS header parameter. It is a claim registered as part of JWT payloads.
`kid` is registered as JWS header parameter but typically used for filtering only.

> We should probably say that `kid` (if DID URL) must correspond to `iss` (if DID) although technically, a VC can have a different `issuer` (DID) and `verificationMethod` (DID) even when using LD-Proofs.

I don't understand how `kid` and `iss` can ever be the same, except in the simplest case where there is only ever a single verification method for a DID.

---

I think from the earlier comments there is some consensus that `kid` is not meant to establish the issuer. It could be abused for it by requiring it to be in the protected header, but it is a stretch and would always lead to discussions simply because the intended semantics don't match what it's used for.

@oed asked for a use case, which I guess could be rephrased as "why not put the issuer in the payload, as is done in JWT?". This kind of works if you restrict the content types of payloads you want to support and assume that all of them have a way to embed a DID issuer. My goal is rather to support arbitrary payloads and establish the issuer at the header level during signature validation / DID resolution, essentially make it a first-class citizen.

**Comment by awoie** [comment]
> > ```
> > * `iss` identifies already the issuer.
> > * `kid` allows you to find the key to verify the JWS/JWT. i agree with @OR13
> > ```
> 
> `iss` is not registered as JWS header parameter. It is a claim registered as part of JWT payloads.

my apologies, i thought the context was JWT-VC. right, `iss` is JWT-specific.

> `kid` is registered as JWS header parameter but typically used for filtering only.
> 
> > We should probably say that `kid` (if DID URL) must correspond to `iss` (if DID) although technically, a VC can have a different `issuer` (DID) and `verificationMethod` (DID) even when using LD-Proofs.
> 
> I don't understand how `kid` and `iss` can ever be the same, except in the simplest case where there is only ever a single verification method for a DID.

nvm, since we are speaking about general JWS. however, for JWT-VCs, `kid` should map onto some `assertionMethod` or `authentication` key in the DID Doc of `iss` or `sub`. otherwise, you could just set any DID for `iss`. 

> 
> I think from the earlier comments there is some consensus that `kid` is not meant to establish the issuer. It could be abused for it by requiring it to be in the protected header, but it is a stretch and would always lead to discussions simply because the intended semantics don't match what it's used for.

If `kid` != (specific DID URL for key of) `iss`, then I'm wondering how anyone could verify the binding between the key/signature and the issuer of the VC.

> 
> @oed asked for a use case, which I guess could be rephrased as "why not put the issuer in the payload, as is done in JWT?". This kind of works if you restrict the content types of payloads you want to support and assume that all of them have a way to embed a DID issuer. My goal is rather to support arbitrary payloads and establish the issuer at the header level during signature validation / DID resolution, essentially make it a first-class citizen.

yes, if you wanted to support anything other than JSON payload, then you won't be able to use `iss`. i guess, this is what LD-Proof folks do with detached JWS (i don't know the internals).

@letmaik what would happen if we'd allow `iss` as a JWS header? Would this have any impact on JWT-VC that defines some mapping for `iss` and `issuer` properties?

**Comment by awoie** [comment]
> I don't understand how `kid` and `iss` can ever be the same, except in the simplest case where there is only ever a single verification method for a DID.

not the same, but `kid` probably needs to be a verification method (key) in the `iss` DID Doc. see other comments above.

**Comment by letmaik** [comment]
> If `kid` != (specific DID URL for key of) `iss`, then I'm wondering how anyone could verify the binding between the key/signature and the issuer of the VC.

VC doesn't define what `kid` should be, since it doesn't have a dependency on DID. It just says "For example, the kid can refer to a key in a DID document, or can be the identifier of a key inside a JWKS. " Whether that's the `id`, or the `kid` of an embedded JWK, or something else, is unspecified. So if you use full DID URLs for `id` in DID documents and decide to use that for `kid`, sure, but the point I'd like to make is that `kid` should not be parsed, but simply matched against some agreed on field, since it is merely used for key selection from the set of keys already established by knowing the issuer and doing key discovery (here via DID resolution of `iss`).

> @letmaik what would happen if we'd allow `iss` as a JWS header? Would this have any impact on JWT-VC that defines some mapping for `iss` and `issuer` properties?

I don't think it has any impact since the mapping in JWT-VC wouldn't change just because a new header parameter is registered. You would have to change the JWT-VC spec to use it, but I don't think there's a point. In the spec that goes along with registration of `iss` as header parameter it could be made clear that the expected usage is meant for JWS profiles that want to be payload agnostic but still say something about the issuer.

**Comment by awoie** [comment]
> but the point I'd like to make is that `kid` should not be parsed, but simply matched against some agreed on field, since it is merely used for key selection from the set of keys already established by knowing the issuer and doing key discovery (here via DID resolution of `iss`).

@letmaik what would be your proposal?

**Comment by letmaik** [comment]
> 
> 
> > but the point I'd like to make is that `kid` should not be parsed, but simply matched against some agreed on field, since it is merely used for key selection from the set of keys already established by knowing the issuer and doing key discovery (here via DID resolution of `iss`).
> 
> @letmaik what would be your proposal?

In my opinion, did-core should not require that `kid` is absent in the JWK. If anything, it should require that it must always be included for widest compatibility with downstream standards.
Each downstream standard (whether JWS or not) that includes key resolution and wants to rely on DID must define how keys are referenced in the DID document.

As an example, for a specific profile of JWS it may be assumed that JWK is used to represent keys and there it makes sense to use the JWK's `kid` and reflect that in the JWS header. This matches the JWS spec most closely: "When used with a JWK, the `kid` value is used to match a JWK `kid` parameter value". If such a profile wants to use DID, then it has to define what the JWS's `kid` value means in relation to the DID document. One option is that it simply says that it references the `publicKeyJwk` that has the given `kid`, while ignoring any verification methods that do not contain `publicKeyJwk`. Another option may be that it matches the ID of a  verification method by constructing ID as `iss#kid`. Yet another option may be that `kid` has to match the ID of the verification method exactly (as suggested in this repo). In the last two options, the actual JWK may be taken from `publicKeyJwk` directly or it could be constructed ad-hoc if a different representation is used, e.g. by converting `publicKeyMultibase` to a JWK. did-core should not be concerned with any of that.

Regarding VC-JWT, I think it was a mistake to not define how DID verification methods should be referenced. I know that VC is independent of DID, but that doesn't preclude it from saying "if the VC issuer is a DID, then kid must be...".
